### PR TITLE
Feat/rs opportunities cv

### PR DESCRIPTION
**Pull request by AlreadyBored**
Add CV MVP functionality for test purposes

### CODE REVIEW

File: client/src/components/Header.tsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { Button, Menu, Dropdown } from 'antd';
 import { GithubAvatar } from 'components/GithubAvatar';
 import css from 'styled-jsx/css';
 
-import { EyeOutlined, EditOutlined, LogoutOutlined, SaveTwoTone } from '@ant-design/icons';
+import { EyeOutlined, EditOutlined, LogoutOutlined, SaveTwoTone /* ExperimentOutlined */ } from '@ant-design/icons';
```
Comment (mikhama) [inline client/src/components/Header.tsx:6]: Do you really need this commented out item?

File: client/src/components/Header.tsx [hunk 2]
```diff
@@ -17,7 +17,12 @@ type Props = {
 };
 
 export function Header(props: Props) {
-  const { isProfilePage, onChangeProfilePageMode, isProfileEditingModeEnabled, isSaveButtonVisible } = props;
+  const {
+    isProfilePage,
+    onChangeProfilePageMode,
+    isProfileEditingModeEnabled,
+    isSaveButtonVisible /* username */,
```
Comment (mikhama) [inline client/src/components/Header.tsx:24]: as well as this one

File: client/src/components/Header.tsx [hunk 3]
```diff
@@ -42,8 +47,17 @@ export function Header(props: Props) {
           <EditOutlined /> Edit
         </Button>
       </Menu.Item>
+      {/*       <Menu.Item key="2" style={isProfileEditingModeEnabled ? menuActiveItemStyle : undefined}>
```
Comment (mikhama) [inline client/src/components/Header.tsx:50]: And this

File: client/src/components/cv/ContactsListCV.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  if (contacts !== null) {
+    const contactsFiltered = Object.entries(contacts);
+    const renderContacts = contactsFiltered.filter((contact): contact is EntryOf<Contacts> => contact[1] !== null);
+    res = (
+      <List
+        size="small"
+        dataSource={renderContacts}
+        renderItem={contact => {
+          const contactType: ContactType = contact[0];
+          const contactText = contact[1];
```
Comment (mikhama) [inline client/src/components/cv/ContactsListCV.tsx:99]: Please use destructuring:
```
const [contactType, contactText] = contact;
```

File: client/src/components/cv/ContactsListCV.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+type EntryOf<T extends object> = { [K in keyof T]: [K, T[K]] }[keyof T];
+
+function ContactsListCV(props: Props) {
+  const { contacts } = props;
+  let res;
+  if (contacts !== null) {
+    const contactsFiltered = Object.entries(contacts);
+    const renderContacts = contactsFiltered.filter((contact): contact is EntryOf<Contacts> => contact[1] !== null);
```
Comment (mikhama) [inline client/src/components/cv/ContactsListCV.tsx:92]: Why do you name it like function?
By it name I think that it's some function that render contacts.

File: client/src/components/cv/ContactsListCV.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  contacts: Contacts | null;
+};
+
+type EntryOf<T extends object> = { [K in keyof T]: [K, T[K]] }[keyof T];
+
+function ContactsListCV(props: Props) {
+  const { contacts } = props;
+  let res;
+  if (contacts !== null) {
+    const contactsFiltered = Object.entries(contacts);
```
Comment (mikhama) [inline client/src/components/cv/ContactsListCV.tsx:91]: I don't see that is filtered.

File: client/src/components/cv/ContactsListCV.tsx (hunk 5 above)
Comment (mikhama) [reply client/src/components/cv/ContactsListCV.tsx:92]: You can rewrite it like this:
```
const omitEmpty = ([, contactText]): contact is EntryOf<Contacts> => contactText !== null);
const contactsFiltered = Object.entries(contacts).filter(omitEmpty);
```

File: client/src/components/cv/EditCV.tsx [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+    userData: null,
+    expires: null,
+  };
+
+  private userFormRef: RefObject<typeof UserDataForm> = React.createRef();
+  private contactsFormRef: RefObject<typeof ContactsForm> = React.createRef();
+
+  private userService = new UserService();
+
+  private nullifyConditional(value: string | null) {
```
Comment (mikhama) [inline client/src/components/cv/EditCV.tsx:46]: private static?

File: client/src/components/cv/EditCV.tsx [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+    await this.setState({
+      contactsList: contactsList,
+      userData: userData,
+      expires: Number(expires),
+    });
+
+    await this.setState({
+      isLoading: false,
+    });
+  }
```
Comment (mikhama) [inline client/src/components/cv/EditCV.tsx:145]: await setState x 2?

File: client/src/components/cv/EditCV.tsx [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    await this.userService.saveCVData(CVData);
+  }
+
+  private async handleSave(data: any) {
+    await this.submitData(data);
+    await this.fetchData();
+  }
+
+  private formatDate(expirationValue: number | null) {
+    if (expirationValue === null || expirationValue === 0) {
```
Comment (mikhama) [inline client/src/components/cv/EditCV.tsx:197]: Interesting check. Need to discuss.

File: client/src/components/cv/EditCV.tsx [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+        <Text>
+          Expiration date: <Text strong>{expirationDateFormatted}</Text>
+        </Text>
+      );
+    }
+  }
+
+  private getDataFromRefs(refs: RefObject<any>[]) {
+    const values = refs
+      .map(ref => {
```
Comment (mikhama) [inline client/src/components/cv/EditCV.tsx:218]: Why not just?:
```
ref => ref.current!.getFieldsValue()
```

File: client/src/components/cv/EditCV.tsx [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+    await this.setState({
+      isLoading: true,
+    });
+    const newExpirationDate = await this.userService.extendCV();
+    await this.setState({
+      expires: Number(newExpirationDate),
+    });
+    await this.setState({
+      isLoading: false,
+    });
```
Comment (mikhama) [inline client/src/components/cv/EditCV.tsx:288]: Why 2 setStates?

File: client/src/pages/cv/index.tsx [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    isLoading: true,
+    editMode: false,
+    opportunitiesConsent: null,
+    errorOccured: false,
+  };
+
+  private userService = new UserService();
+
+  private async switchView(checked: boolean) {
+    if (checked) {
```
Comment (mikhama) [inline client/src/pages/cv/index.tsx:39]: setState({ editMode: !checked })

File: client/src/pages/cv/index.tsx [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    }
+  }
+
+  private async giveConsent(githubId: string) {
+    await this.setState({ isLoading: true });
+    const newConsent = await this.userService.changeOpportunitiesConsent(githubId, true);
+    await this.setState({
+      opportunitiesConsent: newConsent,
+    });
+    await this.setState({ isLoading: false });
```
Comment (mikhama) [inline client/src/pages/cv/index.tsx:56]: setState x 2

File: client/src/pages/cv/index.tsx [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    if (errorOccured) {
+      return <Result status={404} title="User not found" />;
+    }
+
+    if (ownerId === undefined || ownerId instanceof Array) {
+      content = <Result status="warning" title="This page doesn't exist" />;
+    } else {
+      if (isOwner) {
+        if (opportunitiesConsent) {
+          content = (
```
Comment (mikhama) [inline client/src/pages/cv/index.tsx:102]: Seems strange 🤔

File: client/src/pages/employer/index.tsx [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+  };
+
+  private columns = [
+    {
+      title: 'Name',
+      dataIndex: 'complexData',
+      key: 'complexData',
+      render: (data: any) => {
+        const { name, githubId } = data;
+        /*         const { isAdmin } = this.props.session; */
```
Comment (mikhama) [inline client/src/pages/employer/index.tsx:53]: commented code

File: client/src/components/Header.tsx (hunk 1 above)
Comment (mikhama) [reply client/src/components/Header.tsx:6]: //TODO: uncomment after rs-busters tests

File: client/src/components/cv/ViewCV.tsx [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+type State = {
+  isLoading: boolean;
+  contactsList: Contacts | null;
+  userData: UserData | null;
+  coursesData: CourseData[] | null;
+  publicFeedback: PublicFeedback[] | null;
+  expires: number | null;
+};
+
+class ViewCV extends React.Component<Props, State> {
```
Comment (apalchys) [inline client/src/components/cv/ViewCV.tsx:24]: please convert this component to functional component with hooks.
`componentDidMount` could be replaced with 
```typescript
const fetchData = useCallback(async () =>  {
...
}, [...]);

useEffect(() => {
  fetchData();
}, [])
```

File: client/src/components/cv/forms/UserDataForm.tsx [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        </Item>
+        <Item
+          label="Select your English level"
+          wrapperCol={{ span: 24 }}
+          labelCol={{ span: 24 }}
+          name="englishLevel"
+          rules={[{ required: true }]}
+        >
+          <Select placeholder="Not selected yet">
+            <Option value="a0">A0</Option>
```
Comment (apalchys) [inline client/src/components/cv/forms/UserDataForm.tsx:87]: please see `client/src/services/reference-data/english.ts`

File: client/src/pages/cv/index.tsx [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+};
+
+type State = {
+  isLoading: boolean;
+  editMode: boolean;
+  opportunitiesConsent: boolean | null;
+  errorOccured: boolean;
+};
+
+class CVPage extends React.Component<Props, State> {
```
Comment (apalchys) [inline client/src/pages/cv/index.tsx:28]: same here. Please convert it to functional component. Let me know if you need any help.

File: client/src/components/cv/EditCV.tsx [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+      onOk: () => this.props.withdrawConsent(),
+    });
+  }
+
+  private async fetchData() {
+    await this.setState({
+      isLoading: true,
+    });
+
+    const CVData: GetCVData = await this.userService.getCVData(this.props.ownerId);
```
Comment (mikhama) [inline client/src/components/cv/EditCV.tsx:90]: cvData

File: client/src/pages/cv/index.tsx [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+    let content;
+
+    if (errorOccured) {
+      return <Result status={404} title="User not found" />;
+    }
+
+    if (ownerId === undefined || ownerId instanceof Array) {
+      content = <Result status="warning" title="This page doesn't exist" />;
+    } else {
+      if (isOwner) {
```
Comment (mikhama) [inline client/src/pages/cv/index.tsx:100]: if (isOwner) {
   return opportunitiesConsent ? <ViewCV> : <NoConsent>
} else {
   return opportunitiesConsent ? <ViewCV> : <NoConsent>
}

File: server/src/models/user.ts [hunk 21]
```diff
@@ -92,17 +92,68 @@ export class User {
...
   @Column({ nullable: true, type: 'text' })
   militaryService?: string | null;
 
+  @Column({ nullable: true, type: 'text' })
+  cvMilitaryService?: string | null;
+
   @Column({ nullable: true })
   englishLevel?: EnglishLevel;
 
+  @Column({ nullable: true, type: 'text' })
```
Comment (apalchys) [inline server/src/models/user.ts:118]: Please move all `cv..` fields to a separate entity

**Comment by apalchys** [comment]
Closing in favor of https://github.com/rolling-scopes/rsschool-app/pull/696

**Pull request closed without merging**